net/flowtrack/flowtrack.go

```diff
@@ -17,11 +17,38 @@ import (
 	"tailscale.com/types/ipproto"
 )
 
+// MakeTuple makes a Tuple out of netip.AddrPort values.
+func MakeTuple(proto ipproto.Proto, src, dst netip.AddrPort) Tuple {
+	return Tuple{
+		Proto:   proto,
+		Src:     src.Addr().As16(),
+		SrcPort: src.Port(),
+		Dst:     dst.Addr().As16(),
+		DstPort: dst.Port(),
+	}
+}
+
 // Tuple is a 5-tuple of proto, source and destination IP and port.
+//
+// This struct originally used netip.AddrPort, but that was about twice as slow
+// when used as a map key due to the alignment and extra space for the IPv6 zone
+// pointers (unneeded for all our current 2024-06-17 flowtrack needs).
+//
+// This struct is packed optimally and doesn't contain gaps or pointers.
 type Tuple struct {
-	Proto ipproto.Proto  `json:"proto"`
-	Src   netip.AddrPort `json:"src"`
-	Dst   netip.AddrPort `json:"dst"`
+	Src     [16]byte      `json:"src"`
```

Comment:
This could have incorrect overlap for link locals in IPv6, we should add space for the interface index as well

Reply:
Per the comment just above this, we do not.